src/PostgREST/RequestIntent.hs

```diff
@@ -0,0 +1,131 @@
+module PostgREST.RequestIntent where
+
+import qualified Data.Aeson              as JSON
+import qualified Data.ByteString         as BS
+import qualified Data.ByteString.Lazy    as BL
+import qualified Data.Csv                as CSV
+import           Data.List               (find)
+import qualified Data.HashMap.Strict     as M
+import           Data.Maybe              (fromMaybe, isJust, isNothing,
+                                          listToMaybe)
+import           Data.String.Conversions (cs)
+import qualified Data.Text               as T
+import qualified Data.Vector             as V
+import           Network.Wai             (Request (..))
+import           Network.Wai.Parse       (parseHttpAccept)
+import           PostgREST.RangeQuery    (NonnegRange, rangeRequested)
+import           PostgREST.Types         (QualifiedIdentifier (..), Schema)
+
+type RequestBody = BL.ByteString
+
+-- | Types of things a user wants to do to tables/views/procs
+data Action = ActionCreate | ActionRead
+            | ActionUpdate | ActionDelete
+            | ActionInfo   | ActionInvoke
+-- | The target db object of a user action
+data Target = TargetIdent QualifiedIdentifier
+            | TargetRoot
+-- | Enumeration of currently supported content types for
+-- route responses and upload payloads
+data ContentType = ApplicationJSON | TextCSV
+-- | When Hasql supports the COPY command then we can
+-- have a special payload just for CSV, but until
+-- then CSV is converted to a JSON array.
+data Payload = PayloadJSON JSON.Array
+             | PayloadParseError BS.ByteString
+
+-- | Describes what the user wants to do. This data type is a
+-- translation of the raw elements of an HTTP request into domain
+-- specific language.  There is no guarantee that the intent is
+-- sensible, it is up to a later stage of processing to determine
+-- if it is an action we are able to perform.
+data Intent = Intent {
```

Comment:
Should probably have something like
iSelect :: String
iFilters :: [(String, String)]
iOrder :: Maybe String